src/ucp/tag/tag_send.c

```diff
@@ -189,100 +191,136 @@ ucp_tag_send_inline(ucp_ep_h ep, const void *buffer, size_t count,
     }
+    return UCS_INPROGRESS;
+}
 
-    ucp_tag_send_req_init(req, ep, buffer, datatype, count, tag, 0);
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_send_sync_nb,
+                 (ep, buffer, count, datatype, tag, cb),
+                 ucp_ep_h ep, const void *buffer, size_t count,
+                 uintptr_t datatype, ucp_tag_t tag, ucp_send_callback_t cb)
+{
+    ucp_request_param_t param = {
+        .attr_mask = UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FIELD_CALLBACK,
+        .cb.send   = (ucp_send_nbx_callback_t)cb,
+        .datatype  = datatype
+    };
 
-    ret = ucp_tag_send_req(req, count, &ucp_ep_config(ep)->tag.eager,
-                           ucp_ep_config(ep)->tag.rndv.rma_thresh,
-                           ucp_ep_config(ep)->tag.rndv.am_thresh,
-                           cb, ucp_ep_config(ep)->tag.proto, 1);
-out:
-    UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(ep->worker);
-    return ret;
+    return ucp_tag_send_sync_nbx(ep, buffer, count, tag, &param);
 }
 
-UCS_PROFILE_FUNC(ucs_status_t, ucp_tag_send_nbr,
-                 (ep, buffer, count, datatype, tag, request),
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_send_nbx,
+                 (ep, buffer, count, tag, param),
                  ucp_ep_h ep, const void *buffer, size_t count,
-                 uintptr_t datatype, ucp_tag_t tag, void *request)
+                 ucp_tag_t tag, const ucp_request_param_t *param)
 {
-    ucp_request_t *req = (ucp_request_t *)request - 1;
     ucs_status_t status;
+    ucp_request_t *req;
     ucs_status_ptr_t ret;
+    uintptr_t datatype;
 
     UCP_CONTEXT_CHECK_FEATURE_FLAGS(ep->worker->context, UCP_FEATURE_TAG,
-                                    return UCS_ERR_INVALID_PARAM);
+                                    return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM));
     UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
 
-    ucs_trace_req("send_nbr buffer %p count %zu tag %"PRIx64" to %s req %p",
-                  buffer, count, tag, ucp_ep_peer_name(ep), request);
+    ucs_trace_req("send_nbx buffer %p count %zu tag %"PRIx64" to %s",
+                  buffer, count, tag, ucp_ep_peer_name(ep));
 
-    status = UCS_PROFILE_CALL(ucp_tag_send_inline, ep, buffer, count,
-                              datatype, tag);
-    if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
-        UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(ep->worker);
-        return status;
+    datatype = (param->attr_mask & UCP_OP_ATTR_FIELD_DATATYPE) ?
+               param->datatype : ucp_dt_make_contig(1);
+
+    if (ucs_likely(!(param->attr_mask & UCP_OP_ATTR_FLAG_NO_IMM_CMPL))) {
+        status = UCS_PROFILE_CALL(ucp_tag_send_inline, ep, buffer, count,
+                                  datatype, tag);
+        if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
+            ret = UCS_STATUS_PTR(status); /* UCS_OK also goes here */
+            goto out;
+        }
+    }
+
+    if (param->attr_mask & UCP_OP_ATTR_FIELD_REQUEST) {
```

Comment:
I commented before that MPICH only wants to allocate a request if inline send will not succeed. I still do not see a way to avoid *always* allocating/initializing a new request with the new API 😞